src/gt4py/next/embedded/nd_array_field.py

```diff
@@ -148,38 +162,52 @@ def from_array(
         | core_defs.NDArrayObject,  # TODO: NDArrayObject should be part of ArrayLike
         /,
         *,
-        domain: common.Domain,
-        value_type: Optional[type] = None,
+        domain: common.DomainLike,
+        dtype_like: Optional[core_defs.DType] = None,  # TODO define DTypeLike
     ) -> _BaseNdArrayField:
+        domain = common.domain(domain)
         xp = cls.array_ns
-        dtype = None
-        if value_type is not None:
-            dtype = xp.dtype(value_type)
-        array = xp.asarray(data, dtype=dtype)
 
-        value_type = array.dtype.type  # TODO add support for Dimensions as value_type
+        xp_dtype = None if dtype_like is None else xp.dtype(core_defs.dtype(dtype_like).scalar_type)
+        array = xp.asarray(data, dtype=xp_dtype)
+
+        if dtype_like is not None:
+            assert array.dtype.type == core_defs.dtype(dtype_like).scalar_type
 
         assert issubclass(array.dtype.type, core_defs.SCALAR_TYPES)
 
-        assert all(isinstance(d, common.Dimension) for d, r in domain), domain
+        assert all(isinstance(d, common.Dimension) for d in domain.dims), domain
         assert len(domain) == array.ndim
         assert all(
-            len(nr[1]) == s or (s == 1 and nr[1] == common.UnitRange.infinity())
-            for nr, s in zip(domain, array.shape)
+            len(r) == s or (s == 1 and r == common.UnitRange.infinity())
+            for r, s in zip(domain.ranges, array.shape)
         )
 
-        assert value_type is not None  # for mypy
-        return cls(domain, array, value_type)
+        return cls(domain, array)
 
     def remap(self: _BaseNdArrayField, connectivity) -> _BaseNdArrayField:
         raise NotImplementedError()
 
+    def restrict(self, index: common.FieldSlice) -> common.Field | core_defs.ScalarT:
+        new_domain, buffer_slice = self._slice(index)
+
+        new_buffer = self.ndarray[buffer_slice]
+        if len(new_domain) == 0:
+            assert core_defs.is_scalar_type(new_buffer)
+            return new_buffer  # type: ignore[return-value] # I don't think we can express that we return `ScalarT` here
```

Comment:
Should we change the behavior to always return fields as in the DataArray API standard in this PR? What about always requiring `Ellipsis` in relative slicing?

Reply:
No, not in this PR, or we need to have a chat first to figure out how to extract elements (because this PR is preparing itir embedded, which requires it)